### PR TITLE
Agnes distance

### PR DESCRIPTION
**Pull request by targos**

### CODE REVIEW

**Comment by targos** [comment]
Ref: #4

File: .travis.yml [hunk 1]
```diff
@@ -1,4 +1,4 @@
 language: node_js
 node_js:
-  - "0.12"
-  - "iojs"
\ No newline at end of file
+  - "iojs"
+  -"stable"
```
Comment (targos) [inline .travis.yml]: you can remove iojs and instead put:
- 4
- 6
- stable

File: src/agnes.js [hunk 2]
```diff
@@ -153,11 +153,10 @@ function agnes(data, options) {
         dis = 0;
 
     while (list.length > 1) {
-
         // calculates the minimum distance
         d = {};
         min = 10e5;
-        for (var j = 0; j < list.length; j++)
+        for (var j = 0; j < list.length; j++){
```
Comment (targos) [inline src/agnes.js:164]: style: space before the brace

**Comment by targos** [comment]
The commit messages should start with a lowercase. It should also use the imperative tense and the first line must be small. The rest of the explanations go on the next lines. Example:

Instead of

```
Bug: Dependency of ml-stat was removed to keep the package small and to keep compatibility with versions of node.
```

Write something like:

```
fix: remove dependency on ml-stat

This keeps the package small and keeps compatibility with versions of node.
```

**Comment by targos** [comment]
You can change the commit messages with the `git rebase` command

**Comment by andcastillo** [comment]
I commit the messages wrongly again. I'm trying the rebase

**Comment by andcastillo** [comment]
@targos I've finished my changes on the agnes code. The git rebase was cumbersome, and I don't know if I get the correct result. I see new commits but my aim was only to change the commit messages.

File: src/agnes.js [hunk 3]
```diff
@@ -83,43 +73,59 @@ function centroidLink(cluster1, cluster2, disFun) {
...
-    x2 /= cluster2.length;
-    y2 /= cluster2.length;
-    return disFun([x1,y1], [x2,y2])*cluster1.length*cluster2.length / (cluster1.length+cluster2.length);
 }
 
 var defaultOptions = {
     disFunc: euclidean,
-    kind: 'single'
+    kind: 'single',
+    source:'data'
```
Comment (andcastillo) [inline src/agnes.js]: @targos @maasencioh What do you think about the name of the option that specify that the input is a distance matrix??

File: src/agnes.js [hunk 4]
```diff
@@ -83,43 +73,59 @@ function centroidLink(cluster1, cluster2, disFun) {
...
 function agnes(data, options) {
-    options = options || {};
-    for (var o in defaultOptions)
-        if (!(options.hasOwnProperty(o)))
-            options[o] = defaultOptions[o];
+    options = Object.assign({}, defaultOptions, options);
     var len = data.length;
 
+    var distance = data;//If source
+    if(options.source === 'data' ) {
```
Comment (andcastillo) [inline src/agnes.js]: @mljs/collaborators How to deal with the options. In this case, if "source" is set to "data", the input is an array of entries, but if "source" is set to "distance" the input is a distance matrix. But in this case, if source is something different than "data", then it is considered as a distance matrix.  Should I consider the else if(option.source === 'distance') case and throw an error if the options source is not valid??

File: src/agnes.js (hunk 3 above)
Comment (maasencioh) [inline src/agnes.js]: I think that we have two options, to call it `'matrix'` or to put the matrix value in there

Comment (targos) [inline src/agnes.js]: I would do it like in [ml-pca](https://github.com/mljs/pca/blob/5e1b8f361bdba20d9032cb6dab638f741fe8d142/src/pca.js#L44-L47) and have an option like `isDistanceMatrix: true`

**Comment by targos** [comment]
@andcastillo 
I have no idea how you arrived to this result with the rebase (there should not be a merge commit, rebase is kind of the opposite of merge). I restored the branch as it should be and changed the commit messages.
I did a screencast of this, that you can find here: https://asciinema.org/a/d3tglfot8fek4j0mx3k7xme9t

Since I completely rewrote the history of the branch, you need to hard reset it on your side before continuing the work:
- `git checkout agnes-distance`
- `git fetch origin`
- `git reset --hard origin/agnes-distance`

**Comment by targos** [comment]
Here is some information about history rewriting: https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History

**Comment by andcastillo** [comment]
@targos I changed the 'source' option by your suggestion 'isDistanceMatrix'. Can we merge and release a new version of the lib?

**Comment by targos** [comment]
@andcastillo done, thanks for your work

**Pull request merged**